common/flatpak-installation.c

```diff
@@ -991,6 +991,85 @@ _ostree_collection_ref_free0 (OstreeCollectionRef *ref)
   ostree_collection_ref_free (ref);
 }
 
+static gboolean
+flatpak_installation_check_installed_ref_missing_related_ref (FlatpakInstallation *self,
+                                                              const gchar         *remote_name,
+                                                              gchar               *full_ref,
+                                                              GCancellable        *cancellable)
+{
+  g_autoptr(FlatpakDir) dir  = NULL;
+  g_autoptr (GPtrArray) installed_related_refs = NULL;
+  g_autoptr (GPtrArray) remote_related_refs = NULL;
+  g_autoptr (GError) local_error = NULL;
+  guint j, k;
+
+  dir = flatpak_installation_get_dir (self, &local_error);
+  if (dir == NULL)
+    return FALSE;
+
+  installed_related_refs = flatpak_installation_list_installed_related_refs_sync (self,
+                                                                                  remote_name,
+                                                                                  full_ref,
+                                                                                  cancellable,
+                                                                                  &local_error);
+  if (local_error != NULL)
+    {
+      g_warning ("Unable to get installed related refs for %s: %s", full_ref, local_error->message);
+      return FALSE;
+    }
+
+  g_return_val_if_fail (installed_related_refs != NULL, FALSE);
+
+  remote_related_refs = flatpak_installation_list_remote_related_refs_sync (self,
```

Comment:
This creates a lot of FlatpakRef objects etc, but all you need is the ref and should_download, so i recommend using `flatpak_dir_find_remote_related()` directly instead.

Reply:
That way you can also reuse a single FlatpakRemoteState for all the calls for this remotes, avoiding a lot of i/o to re-read it.